src/gui/tagfiltermodel.cpp

```diff
@@ -0,0 +1,326 @@
+        return QModelIndex();
+    return createIndex(row, column, m_tag_items[m_tag_rows.at(row)]);
+}
+
+QModelIndex TagFilterModel::parent(const QModelIndex &) const
+{
+    return QModelIndex();  // Tags are single-level.
+}
+
+int TagFilterModel::rowCount(const QModelIndex &parent) const
+{
+    if (!parent.isValid())
+        return m_tag_rows.count();
+    return 0;
+}
+
+QModelIndex TagFilterModel::index(const QString &tag) const
+{
+    return index(findItem(tag));
+}
+
+QString TagFilterModel::tag(const QModelIndex &index) const
+{
+    if (!index.isValid()) return QString();
+    return static_cast<TagModelItem *>(index.internalPointer())->tag();
+}
+
+QModelIndex TagFilterModel::index(const TagModelItem *item) const
+{
+    const int row = m_tag_rows.indexOf(item->tag());
+    if (row < 0)
+        return QModelIndex();
+    return index(row, 0, QModelIndex());
+}
+
+void TagFilterModel::tagAdded(const QString &tag)
+{
+    const int row = m_tag_rows.count();
+    beginInsertRows(QModelIndex(), row, row);
+    addItemToModel(new TagModelItem(tag, tag));
+    endInsertRows();
+}
+
+void TagFilterModel::tagRemoved(const QString &tag)
+{
+    TagModelItem *item = findItem(tag);
+    if (item) {
+        QModelIndex i = index(item);
+        beginRemoveRows(i.parent(), i.row(), i.row());
+        removeItemFromModel(item);
+        endRemoveRows();
+    }
+}
+
+void TagFilterModel::torrentTagAdded(const QString &tag, int new_tag_count)
+{
+    if (new_tag_count == 1)
+        findItem(getSpecialUntaggedTag())->decreaseTorrentsCount();
+
+    TagModelItem *item = findItem(tag);
+    Q_ASSERT(item);
+
+    item->increaseTorrentsCount();
+    const QModelIndex i = index(item);
+    emit dataChanged(i, i);
+}
+
+void TagFilterModel::torrentTagRemoved(const QString &tag, int new_tag_count)
+{
+    if (new_tag_count <= 0)
+        findItem(getSpecialUntaggedTag())->increaseTorrentsCount();
+
+    TagModelItem *item = findItem(tag);
+    Q_ASSERT(item);
+
+    item->decreaseTorrentsCount();
+    const QModelIndex i = index(item);
+    emit dataChanged(i, i);
+}
+
+void TagFilterModel::torrentAdded(BitTorrent::TorrentHandle *const torrent)
+{
+    findItem(getSpecialAllTag())->increaseTorrentsCount();
+
+    const QVector<TagModelItem *> items = findItems(torrent->tags());
+    if (items.isEmpty())
+        findItem(getSpecialUntaggedTag())->increaseTorrentsCount();
+
+    foreach (TagModelItem *item, items)
+        item->increaseTorrentsCount();
+}
+
+void TagFilterModel::torrentAboutToBeRemoved(BitTorrent::TorrentHandle *const torrent)
+{
+    findItem(getSpecialAllTag())->decreaseTorrentsCount();
+
+    if (torrent->tags().isEmpty())
+        findItem(getSpecialUntaggedTag())->decreaseTorrentsCount();
+
+    foreach (TagModelItem *item, findItems(torrent->tags()))
+        item->decreaseTorrentsCount();
+}
+
+const QString &TagFilterModel::getSpecialAllTag()
+{
+    static const QString *const ALL_TAG = new QString(" ");
```

Comment:
This is a design pattern for safely using objects with static storage duration. Allocating them inside a function ensures thread-safe initialization with well-defined order. By allocating them with new and never destroying them (hence the pointers), you avoid potential race conditions caused by the fact that the C++ compiler only provides weak guarantees on destructor ordering for static objects.

There is no specific issue with this code that requires it, it's just good general practice to avoid bugs being introduced later.

You are right about isValidTag - good catch. It was missing a call to trimmed() on the string to catch whitespace-only tags.